feat(changelog): implement commit processing order as set of steps

dslemusp opened the pull request:
## Description

Currently each commit in the commit list is processed before splitting. This causes an issue when `split_commit` is set to true. Commits in the description are not being processed if the commit message contains a skipped type set in the `commit_parsers`. For instance if the output of a squash and merge start with a `chore:` and the `chore:` is set to skip in the `commit_parsers` the description won't be parsed even if it contains `feat:` or `fix:` or any other non skipped field.

```
// Squash and merge Commit Sample
chore(ci): Update ci runners

fix(ci): a fix --> This is never parsed
chore(ci): another chore --> This is never parsed
```

## Motivation and Context

When working with squash and merge it is useful to have the message of the commit as the overview of the changes and the description as the details of the changes. However it should be possible to skip parsing the message as the description of the commit contains all the info, so trying to avoid repetitions.

## How Has This Been Tested?

I have adjusted the current unit test to reflect the scenario that I am describing

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (no code change)
- [x] Refactor (refactoring production code)
- [ ] Other <!--- (provide information) -->

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] My code follows the code style of this project.
- [ ] I have updated the documentation accordingly.
- [x] I have formatted the code with [rustfmt](https://github.com/rust-lang/rustfmt).
- [x] I checked the lints with [clippy](https://github.com/rust-lang/rust-clippy).
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

<!--- Thank you for contributing to git-cliff! ⛰️  -->

orhun commented:
This was changed in #556 to fix another issue and was reported again in #1093 - can you maybe take a look and double check this?

dslemusp commented:
Yeah actually it seems to me that the initial behaviour and order is what I would expect to have. As I said, right now the commits are treated as one during preprocessing even if the `split_commit` flag is enabled. This results on description being skipped because the commit message is ignored in the preprocessor.

orhun commented:
Thanks for checking. I'd lean towards making this configurable instead of changing the behavior again. We could maybe add a new `processing_order` option like follows:

```toml
[git]
processing_order = ["preprocess", "split", "proces", "postprocess", "etc"]
```

So that you can configure the behavior. What do you think?

ognis1205 commented:
@dslemusp @orhun 

I assume the current step names like `preprocess`, `process`, and `postprocess` are just temporary placeholders for now, since this is still in a draft stage. That said, if we move forward with making the order configurable, I think it’d be good to avoid names like those — they imply a fixed sequence, which might be misleading if the steps can be reordered. It’d be great if we could find names that reflect the actual behavior more clearly.

dslemusp commented:
I agree. I would suggest we use similar (if not same) names as they are defined in the cliff.toml (e.g. `commit_preprocessor` instead of `preprocess` ). I am working on updating the PR but it might take some time!. Feel free to think about the naming in the mean time ;P

orhun commented:
> I would suggest we use similar (if not same) names as they are defined in the cliff.toml

Yes, very good point. I'd like that 👍🏼 

> I am working on updating the PR but it might take some time!

Great, thank you! Let me know if you need any help/guidance in the meantime. I think we should also update the title of this PR accordingly :D

dslemusp commented:
Ok I found some time today I came up with this. Maybe there is still some cleanup to do as this makes https://github.com/orhun/git-cliff/blob/c156fc5edb405737b62a802ed9b1e91f9789ef9b/git-cliff-core/src/commit.rs#L208 not needed.

dslemusp committed:
feat(changelog): implement commit processing order as set of steps

dslemusp committed:
chore: formating

dslemusp commented on git-cliff-core/src/changelog.rs:
```diff
@@ -128,31 +112,141 @@ impl<'a> Changelog<'a> {
+
+    /// Processes the commit list based on the processing order defined in the
+    /// configuration.
+    fn process_commit_list(commits: &mut Vec<Commit<'a>>, git_config: &GitConfig) -> Result<()> {
+        for step in &git_config.processing_order.order {
+            match step {
+                ProcessingStep::CommitParsers => {
+                    debug!("Applying commit parsers...");
+                    *commits = Self::apply_commit_parsers(commits, git_config);
+                }
+                ProcessingStep::CommitPreprocessors => {
+                    debug!("Applying commit preprocessors...");
+                    *commits = Self::apply_commit_preprocessors(commits, git_config);
+                }
+                ProcessingStep::IntoConventional => {
+                    debug!("Converting commits to conventional format...");
+                    *commits = Self::apply_into_conventional(commits, git_config);
+                }
+                ProcessingStep::LinkParsers => {
+                    debug!("Applying link parsers...");
+                    *commits = Self::apply_link_parsers(commits, git_config);
+                }
+                ProcessingStep::SplitCommits => {
+                    debug!("Splitting commits...");
+                    if git_config.split_commits {
+                        *commits = Self::apply_split_commits(commits);
+                    } else {
+                        debug!("Split commits is disabled, skipping...");
+                    }
+                }
+            }
+        }
```
A think to keep in mind is that if all the steps are not listed in the config they will be skipped. So maybe we need to check that the list always contains all the steps (no matter the order)

ognis1205 commented on git-cliff-core/src/config.rs:
```diff
@@ -82,9 +82,54 @@ pub struct ChangelogConfig {
+/// Processing steps for the commits
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub enum ProcessingStep {
```
Suggestions:

1.	Make the enum variant names and config keys consistent and explicit about what the slot does. Prefer either explicit Verb + Object action names or neutral stage names so the intent is clear. Example mappings:

 - `CommitParsers` -> `ParseCommits`
 - `CommitPreprocessors` -> `ModifyCommits`
 - `IntoConventional` -> `ConventionalizeCommits` or `NormalizeCommits`
 - `LinkParsers` -> `ParseLinks`
 - `SplitCommits` -> `SplitCommits`

2.	Rename `CommitPreprocessors` to a name that reflects it being an ordering slot rather than implying it always runs "before" something. Options:

 - If message transformations/modification are the main behavior: `TransformCommits` or `ModifyCommits`.

dslemusp replied:
I think the naming is more a fundamental change as I am using the current naming definitions from the configuration (see https://git-cliff.org/docs/configuration/git). I can indeed modify the names with the caveat of using multiple names for the same (e.g. `ModifyCommits` in place of `CommitPreprocessors` for this step https://git-cliff.org/docs/configuration/git#commit_preprocessors).
It seems that @orhun agreed on using the same names as in the config.toml (https://github.com/orhun/git-cliff/pull/1201#issuecomment-3063303337).

orhun replied:
Yup, I think it makes more sense to use the names that we are already familiar.

ognis1205 reviewed (commented):
Another important point: since there are already issues arising from the interactions of repository and directory-related configs, I think we should be very careful about introducing new functionality into the config. Both the necessity and the exact specification should be carefully evaluated before adding more to this area.

dslemusp committed:
chore: modify processing order to keep original behaviour

dslemusp committed:
debug

dslemusp committed:
chore: lint

dslemusp committed:
chore: refactor tests and reorganize processing mehtods into commit module

dslemusp committed:
chore: fmt

dslemusp commented:
I have rebased and reverted the test to assert the original results. However I cannot manage to make that the parser skip the commit with the footer. I check inside the parser and both in `main` and in my implementation the parser does not skip the commit. Still trying to figure it out. Any help understandig the issue is appreciated

ognis1205 commented on git-cliff-core/src/changelog.rs:
```diff
@@ -90,132 +90,16 @@ impl<'a> Changelog<'a> {
-                        if c.message.is_empty() {
-                            continue;
-                        }
-                        if let Some(c) = Self::process_commit(&c, git_config, summary) {
```
> Still trying to figure it out. Any help understandig the issue is appreciated

I think you're mainly referring to failures in `changelog::test::changelog_generator_split_commits`.

The changes in this PR introduce a breaking change even when the default commit processing order is used.

I haven't traced the logs in full detail, but in the previous implementation, after commit splitting, each split commit message still went through `Commit::process_commit`, which in turn calls things like `Commit::preprocess` and `Commit::into_conventional_commit`.

In the current implementation in this PR, after preprocess is executed, the corresponding processing step doesn't seem to be applied.

This would likely explain the following diff:

```sh
 #### other
 - support unconventional commits
 - this commit is preprocessed
>- use footer
>- footer text
 - make awesome stuff look better
 
 #### ui
 - make good stuff
 
 ### Commit Statistics
 
<- 18 commit(s) contributed to the release.
<- 12 day(s) passed between the first and last commit.
<- 17 commit(s) parsed as conventional.
>- 20 commit(s) contributed to the release.
>- 13 day(s) passed between the first and last commit.
>- 19 commit(s) parsed as conventional.
```

ognis1205 commented on git-cliff-core/src/changelog.rs:
```diff
@@ -1469,6 +1355,13 @@ mod test {
+            ProcessingStep::IntoConventional,
+            ProcessingStep::CommitParsers,
+            ProcessingStep::LinkParsers,
+        ];
```
```suggestion
        config.git.processing_order.order = vec![
            ProcessingStep::CommitPreprocessors,
            ProcessingStep::IntoConventional,
            ProcessingStep::CommitParsers,
            ProcessingStep::SplitCommits,
            ProcessingStep::IntoConventional,
            ProcessingStep::CommitParsers,
            ProcessingStep::LinkParsers,
        ];
```

I've identified the part that needs to be adjusted in order to make the existing unit tests pass.

In the previous implementation (depending on the configuration), the processing flow inside `Changelog::process_commit` -> `Commit::process` was:

`preprocess` -> `into_conventional` -> `parse`

However, in this PR the flow becomes:

`preprocess` -> `parse`

Since `into_conventional` is no longer invoked, the footer regex matcher is not executed. Without calling `into_conventional`, the `footer` is never parsed.

ognis1205 reviewed (changes requested):
With a small adjustment we can make the current tests pass again. However, the existing commit processing order is not strictly linear, it changes depending on configuration and context. Because of that, I'm not yet fully confident that this patch alone is sufficient or entirely correct from a behavioral standpoint.

orhun commented:
#1485
